source/extensions/tracers/xray/localized_sampling.cc

```diff
@@ -0,0 +1,202 @@
+    MessageUtil::loadFromJson(rule_json, document);
+  } catch (EnvoyException& e) {
+    fail("invalid JSON format");
+    return;
+  }
+
+  const auto version_it = document.fields().find(VersionJsonKey);
+  if (version_it == document.fields().end()) {
+    fail("missing version number");
+    return;
+  }
+
+  if (version_it->second.kind_case() != ProtobufWkt::Value::KindCase::kNumberValue ||
+      version_it->second.number_value() != SamplingFileVersion) {
+    fail("wrong version number");
+    return;
+  }
+
+  const auto default_rule_it = document.fields().find(DefaultRuleJsonKey);
+  if (default_rule_it == document.fields().end() ||
+      default_rule_it->second.kind_case() != ProtobufWkt::Value::KindCase::kStructValue) {
+    fail("missing default rule");
+    return;
+  }
+
+  // extract default rule members
+  auto& default_rule_object = default_rule_it->second.struct_value();
+  if (!validateRule(default_rule_object)) {
+    return;
+  }
+
+  default_rule_.setRate(default_rule_object.fields().find(RateJsonKey)->second.number_value());
+  default_rule_.setFixedTarget(static_cast<unsigned>(
+      default_rule_object.fields().find(FixedTargetJsonKey)->second.number_value()));
+
+  const auto custom_rules_it = document.fields().find(CustomRulesJsonKey);
+  if (custom_rules_it == document.fields().end()) {
+    return;
+  }
+
+  if (custom_rules_it->second.kind_case() != ProtobufWkt::Value::KindCase::kListValue) {
+    fail("rules must be JSON array");
+    return;
+  }
+
+  for (auto& el : custom_rules_it->second.list_value().values()) {
+    if (el.kind_case() != ProtobufWkt::Value::KindCase::kStructValue) {
+      fail("rules array must be objects");
+      return;
+    }
+
+    auto& rule_json = el.struct_value();
+    if (!validateRule(rule_json)) {
+      return;
+    }
+
+    LocalizedSamplingRule rule = LocalizedSamplingRule::createDefault();
+    const auto host_it = rule_json.fields().find(HostJsonKey);
+    if (host_it != rule_json.fields().end()) {
+      rule.setHost(host_it->second.string_value());
+    }
+
+    const auto http_method_it = rule_json.fields().find(HttpMethodJsonKey);
+    if (http_method_it != rule_json.fields().end()) {
+      rule.setHttpMethod(http_method_it->second.string_value());
+    }
+
+    const auto url_path_it = rule_json.fields().find(UrlPathJsonKey);
+    if (url_path_it != rule_json.fields().end()) {
+      rule.setUrlPath(url_path_it->second.string_value());
+    }
+
+    // rate and fixed_target must exist because we validated this rule
+    rule.setRate(rule_json.fields().find(RateJsonKey)->second.number_value());
+    rule.setFixedTarget(
+        static_cast<unsigned>(rule_json.fields().find(FixedTargetJsonKey)->second.number_value()));
+
+    custom_rules_.push_back(std::move(rule));
+  }
+}
+
+bool LocalizedSamplingStrategy::shouldTrace(const SamplingRequest& sampling_request) {
+  if (!custom_manifest_.hasCustomRules()) {
+    return shouldTrace(default_manifest_.defaultRule());
+  }
```

Comment:
The way I read this, it's not possible to have a rules JSON that specifies a default fixed target/rate unless there's also at least one custom rule. If I have it right, `default_manifest_.defaultRule()` is always 1/50%. The `custom_manifest_.defaultRule()` may have different values, but unless there's also at least one rule with host/method/url (whether it matches or not) the `custom_manifest_.defaultRule()` is unreachable. But maybe that's by design?

Reply:
That seems strange to me, because it means that to alter those global defaults I have to provide a config like with custom rules when I would expect the following to be sufficient:
```
{
  "version": 2,
  "default": {
    "fixed_target": 1,
    "rate": 0.1
  }
}
```

In the end, it's not Envoy's config format, so I won't belabor this any longer except to ask if a JSON config with no custom rules should be rejected with a validation error?

Reply:
It's unintuitive, I agree. However, It's how all of X-Ray's SDKs work today. I double checked.
>a JSON config with no custom rules should be rejected with a validation error?

I'm reluctant to do that because it would deviate from the behavior of the other X-Ray SDKs.